sw/device/silicon_creator/lib/cert/asn1.c

```diff
@@ -0,0 +1,172 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/cert/asn1.h"
+
+rom_error_t asn1_start(asn1_state_t *new_state, uint8_t *buffer, size_t size) {
+  new_state->buffer = buffer;
+  new_state->size = size;
+  new_state->offset = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_finish(asn1_state_t *state, size_t *out_size) {
+  *out_size = state->offset;
+  state->buffer = 0;
+  state->size = 0;
+  state->offset = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_byte(asn1_state_t *state, uint8_t byte) {
+  return asn1_push_bytes(state, &byte, 1);
+}
+
+rom_error_t asn1_push_bytes(asn1_state_t *state, const uint8_t *bytes,
+                            size_t size) {
+  // TODO: check overflow?
+  if (state->offset + size > state->size)
+    return kErrorAsn1BufferExhausted;
+  memcpy(state->buffer + state->offset, bytes, size);
+  state->offset += size;
+  return kErrorOk;
+}
+
+rom_error_t asn1_start_tag(asn1_state_t *state, asn1_tag_t *new_tag, uint8_t id,
+                           size_t size_hint) {
+  // TODO: use the size hint
+  new_tag->state = state;
+  RETURN_IF_ERROR(asn1_push_byte(state, id));
+  new_tag->len_offset = state->offset;
+  // We do not yet known how many bytes we need to encode the length. For now
+  // reserve three bytes so we can use the 16-bit encoding. This is then fixed
+  // in asn1_finish_tag.
+
+  // Three bytes: one to hold the number of octets to use
+  // and two to store the length.
+  RETURN_IF_ERROR(asn1_push_byte(state, 0x82));
+  RETURN_IF_ERROR(asn1_push_byte(state, 0));
+  RETURN_IF_ERROR(asn1_push_byte(state, 0));
+  new_tag->len_size = 3;
+  return kErrorOk;
+}
+
+rom_error_t asn1_finish_tag(asn1_tag_t *tag) {
+  // Robustness check against accidentely reusing a finished tag.
+  if (tag->state == NULL)
+    return kErrorAsn1Internal;
+  // Compute actually used length.
+  size_t length = tag->state->offset - tag->len_offset - tag->len_size;
+  // Make sure the length is not too large.
+  if (length > 0xffff)
+    return kErrorAsn1Internal;
+  // If the length requires two bytes, then there is no need to move the data.
+  if (length > 0x7f) {
```

Comment:
What is length is in between 0x80 and 0xFF? They should be encoded as `0x81 <length>`, right?

Reply:
yes good catch